File: NtgeCore.Net/test/KeyTest.cs
using System;
using NtgeCore.Net.Ed25519;
using Xunit;

namespace NtgeCore.Net.Test
{
    public class KeyTest
    {
        const string test_publicKey = "pub1pacz8fthpmemnvmehnk6n6y2vcc367j65w8c04xfz9qulhyzw2vqvmy8sg-Ed25519";
        const string test_privateKey = "pri15umhflv9mjwj7spxdjq4chzjacfgryf8hgak9727k0gl2px84kkssvgypc-Ed25519";

        [Fact]
        public void CreateEd25519Keypair()
        {
            using var keypair = Ed25519Keypair.New();
            Assert.True(keypair.Ptr != IntPtr.Zero);
        }

        [Fact]
        public void CreateEd25519PrivateKey()
        {
            using var privateKey = Ed25519PrivateKey.New();
            Assert.True(privateKey.Ptr != IntPtr.Zero);
        }

        [Fact]
        public void SerializeEd25519PrivateKey()
        {
            using var privateKey = Ed25519PrivateKey.New();
            var result = privateKey.Serialize();
            Assert.False(string.IsNullOrEmpty(result));
        }

        [Fact]
        public void SerializeEd25519Keypair()
        {
            using var keypair = Ed25519Keypair.New();
            using var privateKey = keypair.PrivateKey;
            var privateKeyResult = privateKey.Serialize();
            Assert.False(string.IsNullOrEmpty(privateKeyResult));
            using var publicKey = keypair.PublicKey;
            var publicKeyResult = publicKey.Serialize();
            Assert.False(string.IsNullOrEmpty(publicKeyResult));
        }

        [Fact]
        public void DeserializeEd25519PublicKey()
        {
            using var publicKey = Ed25519PublicKey.Deserialize(test_publicKey);
            Assert.True(publicKey.Ptr != IntPtr.Zero);
            Assert.Throws<NtgeException>(() => Ed25519PublicKey.Deserialize(""));
            Assert.Throws<NtgeException>(() => Ed25519PublicKey.Deserialize("hello!"));
            Assert.Throws<NtgeException>(() => Ed25519PublicKey.Deserialize("😉😊😋😎🤣😄😅😗"));
            Assert.Throws<NtgeException>(() => Ed25519PublicKey.Deserialize(test_privateKey));
            Assert.Throws<NtgeException>(() => Ed25519PublicKey.Deserialize("pri15umhflv9mjwj7spxdjq4chzjacfgryf8hgak9727k0gl2px84kkssvgypc-Ed25518"));
        }

        [Fact]
        public void DeserializeEd25519PrivateKey()
        {
            using var privateKey = Ed25519PrivateKey.Deserialize(test_privateKey);
            Assert.True(privateKey.Ptr != IntPtr.Zero);
            Assert.Throws<NtgeException>(() => Ed25519PrivateKey.Deserialize(""));
            Assert.Throws<NtgeException>(() => Ed25519PrivateKey.Deserialize("hello!"));
            Assert.Throws<NtgeException>(() => Ed25519PrivateKey.Deserialize("😉😊😋😎🤣😄😅😗"));
            Assert.Throws<NtgeException>(() => Ed25519PrivateKey.Deserialize(test_publicKey));
            Assert.Throws<NtgeException>(() => Ed25519PrivateKey.Deserialize("pub1pacz8fthpmemnvmehnk6n6y2vcc367j65w8c04xfz9qulhyzw2vqvmy8sg-Ed25518"));
        }

        [Fact]
        public void CreateEd25519KeypairFromEd25519PrivateKey()
        {
            using var privateKey = Ed25519PrivateKey.New();
            using var keypair = Ed25519Keypair.FromPrivateKey(privateKey);
            Assert.True(keypair.Ptr != IntPtr.Zero);
        }

        [Fact]
        public void ConvertEd25519PublicKeyToX25519PublicKey()
        {
            using var publicKey = Ed25519PublicKey.Deserialize(test_publicKey);
            using var x25519Key = publicKey.ToX25519();
            Assert.True(x25519Key.Ptr != IntPtr.Zero);
        }
        [Fact]
        public void ConvertEd25519PrivateKeyToX25519PrivateKey()
        {
            using var privateKey = Ed25519PrivateKey.Deserialize(test_privateKey);
            using var x25519Key = privateKey.ToX25519();
            Assert.True(x25519Key.Ptr != IntPtr.Zero);
        }

        [Fact]
        public void GetEd25519PublicKeyId()
        {
            using var publicKey = Ed25519PublicKey.Deserialize(test_publicKey);
            Assert.True(!string.IsNullOrEmpty(publicKey.KeyId));
        }
    }
}

File: NtgeCore-Android/ntge/src/androidTest/java/com/dimension/ntge/KeyTest.kt
package com.dimension.ntge

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.dimension.ntge.ed25519.Ed25519Keypair
import com.dimension.ntge.ed25519.Ed25519PrivateKey
import com.dimension.ntge.ed25519.Ed25519PublicKey
import org.junit.Assert.*
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class KeyTest {
    @Test
    fun it_create_Ed25519Keypair() {
        Ed25519Keypair.new().use { keypair ->
            assertTrue(keypair.ptr != 0L)
            assertTrue(keypair.publicKey.ptr != 0L)
            assertTrue(keypair.privateKey.ptr != 0L)
        }
    }

    @Test
    fun it_create_Ed25519PrivateKey() {
        Ed25519PrivateKey.new().use { privateKey ->
            assertTrue(privateKey.ptr != 0L)
        }
    }

    @Test
    fun it_serialize_Ed25519PrivateKey() {
        Ed25519PrivateKey.new().use { privateKey ->
            val result = privateKey.serialize()
            assertTrue(result.isNotEmpty())
        }
    }

    @Test
    fun it_serialize_Ed25519Keypair() {
        Ed25519Keypair.new().use { keypair ->
            val publicKey = keypair.publicKey.serialize()
            assertTrue(publicKey.isNotEmpty())
            val privateKey = keypair.privateKey.serialize()
            assertTrue(privateKey.isNotEmpty())
        }
    }

    @Test
    fun it_deserialize_Ed25519PublicKey() {
        Ed25519PublicKey.deserialize(test_publicKey).use { publicKey ->
            assertTrue(publicKey.ptr != 0L)
        }
    }

    @Test
    fun it_deserialize_Ed25519PrivateKey() {
        Ed25519PrivateKey.deserialize(test_privateKey).use { privateKey ->
            assertTrue(privateKey.ptr != 0L)
        }
    }

    @Test
    fun it_should_not_deserialize_Ed25519PublicKey() {
        assertThrows(NtgeException::class.java) {
            Ed25519PublicKey.deserialize("")
        }
        assertThrows(NtgeException::class.java) {
            Ed25519PublicKey.deserialize("hello!")
        }
        assertThrows(NtgeException::class.java) {
            Ed25519PublicKey.deserialize("😉😊😋😎🤣😄😅😗")
        }
        assertThrows(NtgeException::class.java) {
            Ed25519PublicKey.deserialize(test_privateKey)
        }
        assertThrows(NtgeException::class.java) {
            Ed25519PublicKey.deserialize("pri15umhflv9mjwj7spxdjq4chzjacfgryf8hgak9727k0gl2px84kkssvgypc-Ed25518")
        }
    }


    @Test
    fun it_should_not_deserialize_Ed25519PrivateKey() {
        assertThrows(NtgeException::class.java) {
            Ed25519PrivateKey.deserialize("")
        }
        assertThrows(NtgeException::class.java) {
            Ed25519PrivateKey.deserialize("hello!")
        }
        assertThrows(NtgeException::class.java) {
            Ed25519PrivateKey.deserialize("😉😊😋😎🤣😄😅😗")
        }
        assertThrows(NtgeException::class.java) {
            Ed25519PrivateKey.deserialize(test_publicKey)
        }
        assertThrows(NtgeException::class.java) {
            Ed25519PrivateKey.deserialize("pub1pacz8fthpmemnvmehnk6n6y2vcc367j65w8c04xfz9qulhyzw2vqvmy8sg-Ed25518")
        }
    }

    @Test
    fun it_create_Ed25519KeyPair_from_Ed25519PrivateKey() {
        Ed25519PrivateKey.new().use { privateKey ->
            Ed25519Keypair.fromPrivateKey(privateKey).use { keypair ->
                assertTrue(keypair.ptr != 0L)
            }
        }
    }

    @Test
    fun it_convert_Ed25519PublicKey_to_X25519PublicKey() {
        Ed25519PublicKey.deserialize(test_publicKey).use { ed25519PublicKey ->
            ed25519PublicKey.toX25519().use { x25519PublicKey ->
                assertTrue(x25519PublicKey.ptr != 0L)
            }
        }
    }

    @Test
    fun it_convert_Ed25519PrivateKey_to_X25519PrivateKey() {
        Ed25519PrivateKey.new().use { ed25519PrivateKey ->
            ed25519PrivateKey.toX25519().use { x25519PrivateKey ->
                assertTrue(x25519PrivateKey.ptr != 0L)
            }
        }
    }

    @Test
    fun it_get_Ed25519PublicKey_KeyId() {
        Ed25519PublicKey.deserialize(test_publicKey).use { publicKey ->
            assertFalse(publicKey.keyId.isNullOrEmpty())
        }
    }

    @Test
    fun it_sign_and_verify_signature() {
        val message = "Hello, World!".toByteArray(Charsets.UTF_8)
        Ed25519Keypair.new().use { keypair ->
            val signature = keypair.privateKey.sign(message);
            assertTrue(signature.any())
            assertTrue(keypair.publicKey.verify(message, signature))
        }
    }
}

